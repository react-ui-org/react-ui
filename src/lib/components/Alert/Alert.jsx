import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Alert.scss';

export const Alert = ({
  children,
  color,
  icon,
  id,
  onClose,
  translations,
}) => (
  <div
    className={classNames(
      styles.root,
      getRootColorClassName(color, styles),
    )}
    id={id}
    role="alert"
  >
    {icon && (
      <div className={styles.icon}>
        {icon}
      </div>
    )}
    <div
      className={styles.message}
      {...(id && { id: `${id}__content` })}
    >
      {children}
    </div>
    {onClose && (
      <button
        type="button"
        {...(id && { id: `${id}__close` })}
        className={styles.close}
        onClick={() => onClose()}
        onKeyPress={() => onClose()}
        tabIndex="0"
        title={translations.close}
      >
        <span className={styles.closeSign}>Ã—</span>
      </button>
    )}
  </div>
);

Alert.defaultProps = {
  color: 'note',
  icon: null,
  id: undefined,
  onClose: null,
};

Alert.propTypes = {
  /**
   * Alert body.
   */
  children: PropTypes.node.isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the alert.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Optional element to be displayed next to the alert body.
   */
  icon: PropTypes.node,
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__close`
   * * `<ID>__content`
   */
  id: PropTypes.string,
  /**
   * Function to call when the close button is clicked. If not provided, close buttons will be
   * hidden.
   */
  onClose: PropTypes.func,
  /**
   * Translations required by the component.
   */
  translations: PropTypes.shape({
    close: PropTypes.string.isRequired,
  }).isRequired,
};

export const AlertWithContext = withProviderContext(Alert, 'Alert');

export default AlertWithContext;
