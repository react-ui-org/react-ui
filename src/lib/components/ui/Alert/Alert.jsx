import PropTypes from 'prop-types';
import React from 'react';
import styles from './Alert.scss';

const Alert = (props) => {
  let rootTypeClass = styles.note;

  if (props.type) {
    if (props.type === 'success') {
      rootTypeClass = styles.success;
    } else if (props.type === 'error') {
      rootTypeClass = styles.error;
    } else if (props.type === 'warning') {
      rootTypeClass = styles.warning;
    } else if (props.type === 'info') {
      rootTypeClass = styles.info;
    } else if (props.type === 'help') {
      rootTypeClass = styles.help;
    }
  }

  return (
    <div
      className={(`
        ${styles.root}
        ${rootTypeClass}
      `).trim()}
      id={props.id}
      role="alert"
    >
      {props.icon && (
        <div className={styles.icon}>
          {props.icon}
        </div>
      )}
      <div
        className={styles.message}
        {...(props.id && { id: `${props.id}__content` })}
      >
        {props.children}
      </div>
    </div>
  );
};

Alert.defaultProps = {
  icon: null,
  id: undefined,
  type: 'note',
};

Alert.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  icon: PropTypes.element,
  id: PropTypes.string,
  type: PropTypes.oneOf(['error', 'help', 'info', 'note', 'success', 'warning']),
};

export default Alert;
