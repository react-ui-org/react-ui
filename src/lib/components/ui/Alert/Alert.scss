@import 'mixins';

.root {
  @include alert();
}

.icon {
  position: absolute;
  left: $alert-offset;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  height: 100%;
  padding-top: calc((#{$alert-min-height} - (2 * #{$alert-border-width}) - #{$alert-icon-size}) / 2);
}

.message {
  padding-left: $alert-offset;
  font-weight: $alert-message-font-weight;
  line-height: $alert-line-height;

  &:not(:first-child) {
    padding-left: calc(#{$alert-icon-size} + #{$alert-offset});
  }
}

.message strong {
  font-weight: $alert-title-font-weight;
}

.success {
  @include alert-type(success);
}

.warning {
  @include alert-type(warning);
}

.error {
  @include alert-type(error);
}

.info {
  @include alert-type(info);
}

.help {
  @include alert-type(help);
}

.note {
  @include alert-type(note);
}
