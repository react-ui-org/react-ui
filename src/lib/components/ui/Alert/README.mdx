---
name: Alert
menu: 'Components: UI'
route: /components/ui/alert
---

# Alert

import {
  Playground,
  Props,
} from 'docz'
import Icon from '../../../../docs/_components/Icon'
import Button from '../Button'
import {
  AlertWithContext as Alert,
  Alert as ParsableAlert,
} from './Alert'

Alert presents feedback or important information to users.

## Basic Usage

To implement the Alert component, you need to import it first:

```js
import { Alert } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Alert>
    Hello!
  </Alert>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- **Be simple and short** and explain to users why you are interrupting them
  with an alert. Clearly say what happened or what is going to happen. Provide a
  brief title when the message is too long.

- **Use icons** to improve the accessibility of alerts because color may not be
  enough for everyone. [See how](#alerts-with-icons).

## Alert Types

There are six types of alerts (matching all available
[UI colors](/foundation/colors#ui-colors)) to cover all possible needs of your
project: note, success, warning, error, help, and info.

### Note (default)

When no type is specified, a neutral informative alert appears:

<Playground>
  <Alert>
    <strong>Note:</strong> This feature may not be available in all regions.
  </Alert>
</Playground>

### Success

Success alerts confirm that an operation has been processed successfully.

<Playground>
  <Alert type="success">
    <strong>Success:</strong> Settings have been successfully saved.
  </Alert>
</Playground>

### Warning

Use warning alerts when a potentially unfavourable situation may occur. You may
suggest an action to resolve the problem.

<Playground>
  <Alert type="warning">
    <strong>Warning:</strong> Your credit card is going to expire soon.
    {' '}
    <a href="/">Update my payment options</a>
  </Alert>
</Playground>

### Error

Error alerts say there is something that block users from continuing that
requires their immediate attention. The alert should offer a solution to the
problem.

<Playground>
  <Alert type="error">
    <strong>Error:</strong> Cannot connect to the server. Is your internet
    connection working fine?
    {' '}
    <a href="/">Try again</a>
  </Alert>
</Playground>

### Help

This kind of alert can be used to display helpful information.

<Playground>
  <Alert type="help">
    <strong>Help:</strong> You should choose a password you don&apos;t use
    anywhere else.
  </Alert>
</Playground>

### Info

Another common, informative alert.

<Playground>
  <Alert type="info">
    <strong>Info:</strong> This feature depends on user&apos;s OS preferences.
  </Alert>
</Playground>

## Alerts with Icons

An icon can (and should) accompany the message.

ðŸ‘‰ Please note there are no icons pre-packed in React UI. Visit
[Icons](/foundation/icons) to see how it works.

<Playground>
  <Alert type="success" icon={<Icon icon="success" />}>
    <strong>Success:</strong> Settings have been successfully saved.
  </Alert>
</Playground>

## Dismissible Alerts

You can make an alert dismissible by providing a function that closes it on
click on the close button:

<Playground>
  {() => {
    const [isAlertVisible, setIsAlertVisible] = React.useState(true);
    return (
      <>
        {isAlertVisible ? (
          <Alert
            type="success"
            icon={<Icon icon="success" />}
            closeHandler={() => setIsAlertVisible(false)}
          >
            <strong>Success:</strong> Settings have been successfully saved.
          </Alert>
        ) : (
          <Button
            clickHandler={() => setIsAlertVisible(true)}
            label="Show alert"
          />
        )}
      </>
    );
  }}
</Playground>

## API

<Props of={ParsableAlert} />

---

Next: [Badge â†’](/components/ui/badge)
