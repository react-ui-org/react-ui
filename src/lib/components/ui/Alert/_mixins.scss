@import 'variables';

@mixin alert-type($type) {
  $type-properties: map-get($alert-types, $type);
  $foreground-color: map-get($type-properties, foreground-color);

  color: map-get($type-properties, text-color);
  border-color: $foreground-color;
  background-color: map-get($type-properties, background-color);

  a,
  strong,
  .close,
  .icon {
    color: $foreground-color;
  }
}
