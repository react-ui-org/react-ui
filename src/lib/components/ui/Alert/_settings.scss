@use '../../../styles/settings/typography';
@use 'theme';

$font-size: map-get(typography.$size-values, 0);
$line-height: 1.5;
$min-height: calc(#{$font-size} * #{$line-height} + 2 * #{theme.$padding});
