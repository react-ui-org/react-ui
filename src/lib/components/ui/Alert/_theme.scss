$alert-border-width: var(--rui-alert-border-width);
$alert-border-radius: var(--rui-alert-border-radius);
$alert-padding: var(--rui-alert-padding);
$alert-icon-size: var(--rui-icon-size-default);

$alert-types: (
  success: (
    background-color: var(--rui-alert-success-background-color),
    foreground-color: var(--rui-alert-success-foreground-color),
    text-color: var(--rui-alert-success-text-color),
  ),
  warning: (
    background-color: var(--rui-alert-warning-background-color),
    foreground-color: var(--rui-alert-warning-foreground-color),
    text-color: var(--rui-alert-warning-text-color),
  ),
  error: (
    background-color: var(--rui-alert-error-background-color),
    foreground-color: var(--rui-alert-error-foreground-color),
    text-color: var(--rui-alert-error-text-color),
  ),
  info: (
    background-color: var(--rui-alert-info-background-color),
    foreground-color: var(--rui-alert-info-foreground-color),
    text-color: var(--rui-alert-info-text-color),
  ),
  help: (
    background-color: var(--rui-alert-help-background-color),
    foreground-color: var(--rui-alert-help-foreground-color),
    text-color: var(--rui-alert-help-text-color),
  ),
  note: (
    background-color: var(--rui-alert-note-background-color),
    foreground-color: var(--rui-alert-note-foreground-color),
    text-color: var(--rui-alert-note-text-color),
  ),
);
