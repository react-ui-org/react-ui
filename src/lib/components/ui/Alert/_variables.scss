@import '../../../styles/settings/typography';
@import './theme';

$alert-font-size: map-get($typography-size-values, 0);
$alert-line-height: 1.5;
$alert-min-height: calc(#{$alert-font-size} * #{$alert-line-height} + 2 * #{$alert-padding});
$alert-stripe-width: 5px;
$alert-message-font-weight: map-get($typography-font-weight-values, light);
$alert-title-font-weight: map-get($typography-font-weight-values, bold);
