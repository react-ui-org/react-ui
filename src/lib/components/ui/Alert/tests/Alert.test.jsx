import React from 'react';
import {
  shallow,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Icon from '../../Icon';
import Alert from '../Alert';

jest.mock('../../Icon/load-material-design-icons');

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(
      <Alert
        message="Message"
      >
        <div>Children</div>
      </Alert>,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(
      <Alert
        icon={<Icon icon="info" />}
        id="custom-id"
        message="Message"
        type="success"
      >
        <div>Children</div>
      </Alert>,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
