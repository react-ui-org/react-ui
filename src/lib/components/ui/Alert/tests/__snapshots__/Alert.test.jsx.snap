// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root
        note"
  role="alert"
>
  <div
    className="message"
  >
    <div>
      Children
    </div>
  </div>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
        success"
  id="custom-id"
  role="alert"
>
  <div
    className="icon"
  >
    <Icon
      icon="info"
      size="medium"
    />
  </div>
  <div
    className="message"
    id="custom-id__content"
  >
    <div>
      Children
    </div>
  </div>
</div>
`;
